stdlib/mexpr/parser.mc

```diff
@@ -0,0 +1,433 @@
+  utest eatWSAC (initPos "") " /- foo -/ bar" with {str = "bar", pos = posVal "" 1 11} in
+  utest eatWSAC (initPos "") " /- foo\n x \n -/ \nbar " with {str = "bar ", pos = posVal "" 4 0} in
+  utest eatWSAC (initPos "") " /- x -- y /- foo \n -/ -/ !" with {str = "!", pos = posVal "" 2 7} in
+  ()
+
+
+
+-- Top of the expression parser. Connects WSAC with parsing of other non terminals
+lang ExprParser = WSACParser
+  sem parseExpr (p: Pos) =
+  | s ->
+    let r1 = parseExprMain p 0 s in
+    let r2 = eatWSAC r1.pos r1.str in
+    if eqi (length r2.str) 0 then r1.val
+    else posErrorExit r2.pos "Parse error. Unknown characters."
+  
+  sem parseExprMain (p: Pos) (prec: Int) =
+  | s ->
+    let r1 = eatWSAC p s in
+    let exp = parseExprImp r1.pos r1.str in
+    let r2 = eatWSAC exp.pos exp.str in
+    parseInfix r2.pos prec exp r2.str 
+
+  sem parseInfix (p: Pos) (prec: Int) (exp: Expr) =
+
+  sem parseExprImp (p: Pos) =
+  | _ -> posErrorExit p "Parse error. Unknown character sequence."
+end
+
+
+-- Include this fragment is there are no infix operations
+lang ExprParserNoInfix = ExprParser
+  sem parseInfix (p: Pos) (prec: Int) (exp: Expr) =
+  | _ -> exp
+end
+
+-- Parsing of boolean literals
+lang BoolParser = ExprParser + ConstAst + BoolAst
+  sem parseExprImp (p: Pos) =
+  | "true" ++ xs ->
+      let p2 = advanceCol p 4 in
+      {val = TmConst {val = CBool {val = true}, fi = makeInfo p p2}, pos = p2, str = xs}
+  | "false" ++ xs ->
+      let p2 = advanceCol p 5 in
+      {val = TmConst {val = CBool {val = false}, fi = makeInfo p p2}, pos = p2, str = xs}
+end
+
+
+-- Parsing of an unsigned integer
+lang UIntParser = ExprParser + ConstAst + IntAst
+  sem parseExprImp (p : Pos) =
+  | ("0" ++ s) & xs | ("1" ++ s) & xs | ("2" ++ s) & xs | ("3" ++ s) & xs | ("4" ++ s) & xs |
+    ("5" ++ s) & xs | ("6" ++ s) & xs | ("7" ++ s) & xs | ("8" ++ s) & xs | ("9" ++ s) & xs ->
+    recursive
+    let work = lam p2. lam str. lam num.
+      match str with [x] ++ xs then
+        let c = char2int x in
+        if and (geqi c 48) (leqi c 57)
+        then work (advanceCol p2 1) xs (snoc num x)
+        else {val = TmConst {val = CInt {val = string2int num}, fi = makeInfo p p2}, pos = p2, str = str}
+      else {val = TmConst {val = CInt {val = string2int num}, fi = makeInfo p p2}, pos = p2, str = str}
+    in work (advanceCol p 1) s ([head xs])
+end
+
+
+-- Fragment for simple parsing of keyword
+lang KeywordParser = WSACParser
+  sem matchKeyword (keyword: String) (p: Pos) =
+  | s ->
+     let r = eatWSAC p s in
+     if isPrefix eqc keyword r.str then
+       let l = length keyword in
+       {pos = advanceCol r.pos l, str = slice r.str l (subi (length r.str) l)}
+     else
+       posErrorExit r.pos (join ["Unknown character. Expected '", keyword, "'."])
+end
+
+
+-- Parsing if expressions
+lang IfParser = ExprParser + KeywordParser +  MatchAst + BoolPat
+  sem parseExprImp (p: Pos) =
+  | "if" ++ xs ->
```

Comment:
Does this handle parsing `"iftrue 4"` properly (note the missing space, it's a function application of the function `iftrue` to `4`)? I feel like it will trigger on any identifier beginning with `if`.

Reply:
Good point. It is now fixed.